### PR TITLE
Implement Seamock Library

### PR DESCRIPTION
**Pull request by AdrianJendo**

### CODE REVIEW

File: src/seamock-lib/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+            }
+        }
+
+        let times_attr = Ident::new(&format!("times_{}", &method.sig.ident), method.sig.ident.span());
+        let max_times_attr = Ident::new(&format!("max_times_{}", &method.sig.ident), method.sig.ident.span());
+        let val_attr = Ident::new(&format!("val_with_{}", &method.sig.ident), method.sig.ident.span());
+        let ret_func = Ident::new(&format!("val_returning_{}", &method.sig.ident), method.sig.ident.span());
+        let method_name_string =  &method.sig.ident.to_string();
+
+        // TODO: generate code to check parameter values are correct
```
Comment (priyasiddharth) [inline src/seamock-lib/src/lib.rs]: can this be added as a unit test?

File: src/seamock-lib/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+    let expanded = quote! {
+        use core::cell::RefCell;
+        #input
+        #mock_struct
+        #mock_impl
+    };
+
+
+    TokenStream::from(expanded)
+}
```
Comment (priyasiddharth) [inline src/seamock-lib/src/lib.rs]: nit:add newline at EOF

File: src/rust-jobs/mock-test1/lib.rs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+#[no_mangle]
+pub extern "C" fn entrypt() {
+    let mut x: MockTest = MockTest::new();
+	let mut y: i32 = verifier::any!();
+    let mut z: bool = verifier::any!();
+
+    verifier::assume!(y < 10);
+
+    x
```
Comment (priyasiddharth) [inline src/rust-jobs/mock-test1/lib.rs]: typo?

File: src/rust-jobs/mock-test1/lib.rs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        .times_a(2)
+        .returning_a(|z, a| a + 6)
+        .returning_b(|| 4)
+        .times_b(2);
+    
+    // verifier::vassert!(y == x.c());
```
Comment (priyasiddharth) [inline src/rust-jobs/mock-test1/lib.rs]: why comment it out... remove if unneeded

File: scripts/rust-jobs.json [hunk 5]
```diff
@@ -55,4 +56,4 @@
         "vec-filter",
         "vec-sort-reverse"
     ]
-}
+}
```
Comment (priyasiddharth) [inline scripts/rust-jobs.json]: nit:newline at EOF

**Review by priyasiddharth — COMMENTED** [review]
Minor comments.

File: src/seamock-lib/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+        let times_attr = Ident::new(&format!("times_{}", &method.sig.ident), method.sig.ident.span());
+        let max_times_attr = Ident::new(&format!("max_times_{}", &method.sig.ident), method.sig.ident.span());
+        let val_attr = Ident::new(&format!("val_with_{}", &method.sig.ident), method.sig.ident.span());
+        let ret_func = Ident::new(&format!("val_returning_{}", &method.sig.ident), method.sig.ident.span());
+        let method_name_string =  &method.sig.ident.to_string();
+
+        Some (quote! {
+            fn #method_name(#method_inputs) #method_output {
+                self.#times_attr.replace_with(|&mut old| old + 1);
+                // verifier::vassert!(*self.#times_attr.borrow() <= self.#max_times_attr)
```
Comment (AdrianJendo) [inline src/seamock-lib/src/lib.rs]: This was not working, any suggestions on how to handle it?

**Pull request merged**